ENIGMAsystem/SHELL/Audio_Systems/DirectSound/DSsystem.cpp

```diff
@@ -207,15 +205,16 @@ int sound_add_from_buffer(int id, void* buffer, size_t bufsize) {
         bits = waveHeader->bitsPerSample,
         size = waveHeader->dataSize;
 
-  snd->length = size / channels / (bits / 8) / freq;
+  snd.length = size / channels / (bits / 8) / freq;
 
   delete waveHeader;
 
-  snd->soundBuffer->SetCurrentPosition(0);
+  snd.soundBuffer->SetCurrentPosition(0);
   // Set volume of the buffer to 100%.
-  snd->soundBuffer->SetVolume(0);
-  snd->loaded = LOADSTATE_COMPLETE;
+  snd.soundBuffer->SetVolume(0);
+  snd.loaded = LOADSTATE_COMPLETE;
 
+  sound_resources.assign(id, snd);
```

Comment:
How is this copy performed? How much data is copied? Should you change the class to require a move constructor?

Reply:
This is the final piece of the puzzle to work out, the move and copy construction for deep objects. Sound just happens to be one of those resources since it holds a pointer to the DirectSound buffer.

Reply:
We shouldn't leave this a mystery while we're replacing storage with it. Adopting move semantics is a fine way to prevent copying large chunks of memory around.

Reply:
Alright, I think we've worked out the move and copy semantics now. Give it one last look over.